components/backup_text/src/rwer.rs

```diff
@@ -19,13 +22,60 @@ use tikv_util::error;
 use tipb::{ColumnInfo, TableInfo};
 use txn_types::Key;
 
+trait CountedWrite: Write {
+    fn get_size(&self) -> u64;
+}
+
+struct CountedWriter<W> {
+    file_size: u64,
+    writer: W,
+}
+
+impl<W: Write> CountedWriter<W> {
+    fn new(writer: W) -> Self {
+        Self {
+            file_size: 0,
+            writer,
+        }
+    }
+}
+
+impl<W: Write> Write for CountedWriter<W> {
+    #[inline]
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        let result = self.writer.write(buf);
+        if let Ok(size) = result {
+            self.file_size += size as u64;
+        }
+        result
+    }
+
+    #[inline]
+    fn flush(&mut self) -> io::Result<()> {
+        self.writer.flush()
+    }
+}
+
+impl<W: Write> CountedWrite for CountedWriter<W> {
+    #[inline]
+    fn get_size(&self) -> u64 {
+        self.file_size
+    }
+}
+
+impl<W: Write> CountedWrite for ZlibEncoder<W> {
+    #[inline]
+    fn get_size(&self) -> u64 {
+        self.total_out()
+    }
+}
+
 pub struct TextWriter {
     ctx: EvalContext,
     data_type: Option<DataType>,
     format: FileFormat,
-    file_writer: BufWriter<File>,
+    file_writer: Box<dyn CountedWrite>,
```

Comment:
How about `file_writer: BufWriter<Box<dyn Write>>`?

Reply:
We can get the file size by `fs::metadata` after flush.

Reply:
If so, we must `flush` before checking the metadata for the file size, which may lead to `get_size()` requiring a mutable reference to `self`.

Reply:
> If so, we must flush before checking the metadata for the file size, which may lead to get_size() requiring a mutable reference to self.

There are only a few use cases of `get_size`, and all of them are called after `flush`, we can store the file size after flush, and `get_size(&self) -> Option<usize>`

> Also, considering that ZlibEncoder requires an additional finish() step, while currently TextWriter::finish won't drop (and finish) the file_writer, we may still need to add this trait for different writers.

I think we don't need that  additional `ZlibEncoder::finish` step, which is nearly the same of `<ZlibEncoder as Write>::flush`, see:
https://docs.rs/flate2/1.0.22/src/flate2/zlib/write.rs.html#160-162
https://docs.rs/flate2/1.0.22/src/flate2/zlib/write.rs.html#137-140

Reply:
`flush_finish` differs from `finish`, since it is used only if the underlying `Write` is reused for encoding `zlib` data in the future. In other words, `flush_finish` won't write the ending bytes.

Reply:
When `ZlibEncoder` is dropped `finish` will be called or if this is too hack, we can add a wrapper for `ZlibEncoder` and implement:
```
struct Wrappper(ZlibEncoder<File>);
impl Write for Wrappper {
    fn flush(&mut self) -> io::Result<()> {
        self.0.flush()?;
        self.0.try_finish()?;
        Ok(())
    }
    ...
}
``` 
still, I think this is cleaner than `CountedWriter`

Reply:
`try_finish` can only be called once if there's no more `write` after that. 🤨

Reply:
Yes, and `file_writer.flush` is only call in `TextWriter::finish` and `TextWriter::finish_read`.

Reply:
This is only one of the problems. Without a trait we may also have no way to know how many bytes are indeed written to the file. Due the possible existence of `Encoder`, the counts of in-bytes and out-bytes are different.

Reply:
This is only one of the problems. Without a trait we may also have no way to know how many bytes are indeed written to the file. Due the possible existence of `Encoder`, the counts of in-bytes and out-bytes are different.

Reply:
> We can get the file size by fs::metadata after flush.

I think this could work